### PR TITLE
checker(dm): support optimistic checker

### LINKED ISSUE #4328: Add optimistic compitible table structure check

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve?

Issue Number: close #4328

### What is changed and how it works?

- don't check sharding tables if exist checkpoint
- support to check optimistic compitible

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update key monitor metrics in both TiCDC document and official document

### Release note 

```release-note
Please add a release note.
If you don't think this PR needs a release note then fill it with `None`.
```

### CODE REVIEW

File: dm/checker/checker.go
Comment (Ehco1996) [inline dm/checker/checker.go:258]: how about `utils.GetTableCreateSQL` instead?

File: dm/checker/checker.go
Comment (Ehco1996) [inline dm/checker/checker.go:268]: seems you forget the `checkingShard` options

File: dm/pkg/checker/table_structure.go
Comment (Ehco1996) [inline dm/pkg/checker/table_structure.go:581]:
```suggestion
		Desc:  "check consistency of sharding table structures for Optimistic Sharding Merge",
```

**Review by Ehco1996 — COMMENTED** [review]
rest LGTM

File: dm/checker/checker.go
Comment (okJiang) [reply dm/checker/checker.go:258]: I think they are no different. So it is ok to stay it.

File: dm/checker/checker.go
Comment (okJiang) [reply dm/checker/checker.go:268]: `checkingShard` is in L249

**Comment by lance6716** [comment]
> don't check sharding tables if not exist checkpoint

this PR description is wrong?

File: dm/checker/check_test.go
Comment (lance6716) [inline dm/checker/check_test.go:313]: is there a test to cover checkpoint existing?

**Comment by okJiang** [comment]
> > don't check sharding tables if not exist checkpoint
> 
> this PR description is wrong?

my bad. The correct is "don't check sharding tables if exist checkpoint"

File: dm/checker/checker.go
Comment (lance6716) [inline dm/checker/checker.go:252]: I forget if we should check shard table structure for pessimistic incremental task 😭 currently your code will skip adding checker for pessimistic incremental task . Could you left some comment of decision?

File: dm/pkg/checker/table_structure.go
Comment (lance6716) [inline dm/pkg/checker/table_structure.go:555]:
```suggestion
// * check whether they have compatible column list
```

maybe

File: dm/pkg/checker/table_structure.go
Comment (lance6716) [inline dm/pkg/checker/table_structure.go:574]:
```suggestion
	return fmt.Sprintf("optimistic sharding table %s consistency checking", c.targetTableID)
```

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: dm/checker/checker.go
Comment (Ehco1996) [inline dm/checker/checker.go:254]: seem this should impl in  `func (c *Checker) IsFreshTask() (bool, error) `

light suggestion, maybe do this job in next pr or left o todo item here

**Review by Ehco1996 — COMMENTED** [review]
LGTM after all comments are resloved

File: dm/checker/checker.go
Comment (okJiang) [reply dm/checker/checker.go:252]: https://github.com/pingcap/tiflow/blob/e2d529cc9887ab61063555dbbbf1d9f9765581e3/dm/docs/RFCS/20211130_enhanced_pre_checker.md?plain=1#L53

File: dm/checker/check_test.go
Comment (okJiang) [reply dm/checker/check_test.go:313]: it isn't. These code mock 

```go
		checkpointSQLs := []string{
			fmt.Sprintf("SHOW CREATE TABLE %s", dbutil.TableName(instance.cfg.MetaSchema, cputil.LoaderCheckpoint(instance.cfg.Name))),
			fmt.Sprintf("SHOW CREATE TABLE %s", dbutil.TableName(instance.cfg.MetaSchema, cputil.LightningCheckpoint(instance.cfg.Name))),
			fmt.Sprintf("SHOW CREATE TABLE %s", dbutil.TableName(instance.cfg.MetaSchema, cputil.SyncerCheckpoint(instance.cfg.Name))),
		}
		var existCheckpoint bool
		for _, sql := range checkpointSQLs {
			c.tctx.Logger.Info("exec query", zap.String("sql", sql))
			_, err := instance.targetDB.DB.QueryContext(c.tctx.Ctx, sql)
			if err != nil && !utils.IsMySQLError(err, mysql.ErrNoSuchTable) {
				return err
			}
			if err == nil {
				existCheckpoint = true
				c.tctx.Logger.Info("exist checkpoint, so don't check sharding tables")
				break
			}
		}
```

**Comment by okJiang** [comment]
all comment fixed, ptal again, thank you @lance6716 @Ehco1996 

/run-dm-integration-test

**Review by Ehco1996 — APPROVED** [review]
LGTM~~

**Comment by okJiang** [comment]
/hold 
change optimistic-checker to multi-thread.

**Review by lance6716 — APPROVED** [review]

**Pull request merged**
